src/client/java/daos-java-api/src/main/native/com_intel_daos_client_DaosFsClient.c

```diff
@@ -0,0 +1,911 @@
+	memcpy(&poh, &poolPtr, sizeof(poh));
+	daos_cont_info_t co_info;
+	const char *cont_str = (*env)->GetStringUTFChars(env, contUuid, NULL);
+	uuid_t cont_uuid;
+	uuid_parse(cont_str, cont_uuid);
+	daos_handle_t coh;
+	jlong ret = -1;
+	int rc = daos_cont_open(poh, cont_uuid, mode, &coh, &co_info, NULL);
+	if (rc) {
+		char *tmp = "Failed to open container (id: %s)";
+		char *msg = (char *)malloc(strlen(tmp) + strlen(cont_str));
+		sprintf(msg, tmp, cont_str);
+		throw_exception_base(env, msg, rc, 1, 0);
+		ret = -1;
+	} else {
+		memcpy(&ret, &coh, sizeof(coh));
+	}
+	(*env)->ReleaseStringUTFChars(env, contUuid, cont_str);
+	return ret;
+}
+
+JNIEXPORT void JNICALL Java_com_intel_daos_client_DaosFsClient_daosCloseContainer
+  (JNIEnv *env, jclass clientClass, jlong contPtr){
+	daos_handle_t coh;
+	memcpy(&coh, &contPtr, sizeof(coh));
+	int rc = daos_cont_close(coh, NULL);
+	if(rc){
+		printf("Failed to close container rc: %d\n", rc);
+		printf("error msg: %s\n", d_errstr(rc));
+	}
+}
+
+JNIEXPORT jlong JNICALL Java_com_intel_daos_client_DaosFsClient_dfsMountFs
+  (JNIEnv *env, jclass clientClass, jlong poolPtr, jlong contPtr, jboolean readOnly){
+	int flags = readOnly ? O_RDONLY : O_RDWR;
+	dfs_t *dfsPtr;
+	daos_handle_t poh;
+	daos_handle_t coh;
+	memcpy(&poh, &poolPtr, sizeof(poh));
+	memcpy(&coh, &contPtr, sizeof(coh));
+	int rc = dfs_mount(poh, coh, flags, &dfsPtr);
+	if(rc){
+		char *msg = "Failed to mount fs ";
+		throw_exception_const_msg(env, msg, rc);
+		return -1;
+	}
+	return *(jlong*)&dfsPtr;
+}
+
+JNIEXPORT void JNICALL Java_com_intel_daos_client_DaosFsClient_dfsUnmountFs
+  (JNIEnv *env, jclass clientClass, jlong dfsPtr){
+	dfs_t *dfs = *(dfs_t**)&dfsPtr;
+	int rc = dfs_umount(dfs);
+	if(rc){
+		printf("Failed to unmount fs rc: %d\n", rc);
+		printf("error msg: %s\n", strerror(rc));
+	}
+}
+
+JNIEXPORT jlong JNICALL Java_com_intel_daos_client_DaosFsClient_dfsMountFsOnRoot
+  (JNIEnv *env, jclass clientClass, jlong poolPtr){
+	dfs_t *dfsPtr;
+	daos_handle_t poh;
+	memcpy(&poh, &poolPtr, sizeof(poh));
+	int rc = dfs_mount_root_cont(poh, &dfsPtr);
+	if(rc){
+		char *msg = "Failed to mount fs on root container";
+		throw_exception_const_msg(env, msg, rc);
+		return -1;
+	}
+	return *(jlong*)&dfsPtr;
+}
+
+JNIEXPORT void JNICALL Java_com_intel_daos_client_DaosFsClient_dfsUnmountFsOnRoot
+  (JNIEnv *env, jclass clientClass, jlong dfsPtr){
+	dfs_t *dfs = *(dfs_t**)&dfsPtr;
+	int rc = dfs_umount_root_cont(dfs);
+	if(rc){
+		printf("Failed to unmount fs on root container rc: %d\n", rc);
+		printf("error msg: %s\n", strerror(rc));
+	}
+}
+
+JNIEXPORT void JNICALL Java_com_intel_daos_client_DaosFsClient_daosFinalize
+  (JNIEnv *env, jclass clientClass){
+	int rc = daos_fini();
+	if(rc){
+		printf("Failed to finalize daos rc: %d\n", rc);
+		printf("error msg: %s\n", d_errstr(rc));
+	}
+}
+
+JNIEXPORT void JNICALL Java_com_intel_daos_client_DaosFsClient_move
+  (JNIEnv *env, jobject obj, jlong dfsPtr, jstring srcPath, jstring destPath){
```

Comment:
this comment applies to all the DFS bindings that you have:
the reason we don't do full path for source and destination is because you will need to do a lookup each time like you are doing below. and this is bad for performance.
the application can do a lookup once and use the parent multiple times as needed without having to do lookup each time. any reason why you don't mirror that through the java API and you use full path each call?

Reply:
DaosFsClient is low level class in Java wrapper. If you look at my java class, DaosFile, you will see that it tries to remember object id of opened fs object for multiple fs operations on same file. As you can see, only user knows when and where to use cached fs object. In most cases, we just move file without need to open its parent. 

But your comments do remind me to add another move method with opened parent as parameter just in case.

Reply:
ah i see, thx for pointing that out